pkg/devfile/adapters/common/utils.go

```diff
@@ -20,3 +30,158 @@ func GetSupportedComponents(data versions.DevfileData) []common.DevfileComponent
+	}
+
+	if action.Component == nil || *action.Component == "" {
+		return false
+	}
+
+	if action.Command == nil || *action.Command == "" {
+		return false
+	}
+
+	return true
+}
+
+// GetBuildCommand iterates through the components in the devfile and returns the build command
+func GetBuildCommand(data versions.DevfileData, devfileBuildCmd string) (buildCommand common.DevfileCommand) {
+	if devfileBuildCmd != "" {
+		buildCommand = GetCommand(data, devfileBuildCmd)
+	} else {
+		buildCommand = GetCommand(data, string(DefaultDevfileBuildCommand))
+	}
+
+	return
+}
+
+// GetRunCommand iterates through the components in the devfile and returns the run command
+func GetRunCommand(data versions.DevfileData, devfileRunCmd string) (runCommand common.DevfileCommand) {
+	if devfileRunCmd != "" {
+		runCommand = GetCommand(data, devfileRunCmd)
+	} else {
+		runCommand = GetCommand(data, string(DefaultDevfileRunCommand))
+	}
+
+	return
+}
+
+// IsCommandPresent checks if the given command is empty or not
+func IsCommandPresent(command common.DevfileCommand) bool {
+	var emptyCommand common.DevfileCommand
+	isPresent := false
+
+	if !reflect.DeepEqual(emptyCommand, command) {
+		isPresent = true
+	}
+
+	return isPresent
+}
+
+// ValidateAndGetPushDevfileCommands validates the build and the run command,
+// if provided through odo push or else checks the devfile for devBuild and devRun.
+// It returns the build and run commands if its validated successfully, error otherwise.
+func ValidateAndGetPushDevfileCommands(data versions.DevfileData, devfileBuildCmd, devfileRunCmd string) ([]common.DevfileCommand, error) {
+	var pushDevfileCommands []common.DevfileCommand
+	// var emptyCommand common.DevfileCommand
+	validateBuildCommand, validateRunCommand := false, false
+
+	buildCommand := GetBuildCommand(data, devfileBuildCmd)
+	if devfileBuildCmd == "" && !IsCommandPresent(buildCommand) {
+		// If there is no build command either in the devfile or through odo push, default validate to true since build command is optional
+		validateBuildCommand = true
+		glog.V(3).Infof("No Build command was provided")
+	} else if IsCommandPresent(buildCommand) && IsCommandValid(data, buildCommand) {
+		// If the build command is present, validate it
+		pushDevfileCommands = append(pushDevfileCommands, buildCommand)
+		validateBuildCommand = true
+		glog.V(3).Infof("Build command %v validated", buildCommand.Name)
+	}
+
+	runCommand := GetRunCommand(data, devfileRunCmd)
+	if IsCommandPresent(runCommand) && IsCommandValid(data, runCommand) {
+		// If the run command is present, validate it
+		pushDevfileCommands = append(pushDevfileCommands, runCommand)
+		validateRunCommand = true
+		glog.V(3).Infof("Run command %v validated", runCommand.Name)
+	}
+
+	if !validateBuildCommand || !validateRunCommand {
+		return []common.DevfileCommand{}, fmt.Errorf("devfile command validation failed, validateBuildCommand: %v validateRunCommand: %v", validateBuildCommand, validateRunCommand)
```

Comment:
This is how the error currently looks for the user:
```
✗  Failed to start component with name openlibertydevfile.
Error: Failed to create the component: failed to validate devfile build and run commands: devfile command validation failed, validateBuildCommand: true validateRunCommand: false
```
With the changes to `validateAction()` we can just print each of the error messages.

Reply:
I added commit https://github.com/openshift/odo/commit/5378ed9f349421c54204ede014ed827ba4b29e66 to address this